### PR TITLE
Fix timeline

### LINKED ISSUE #165: Timeline does not work properly
### Current Behavior

The first row of the timeline shows all possible objects with different dates

### Steps To Reproduce

1. go to [https://forschung.stadtgeschichtebasel.ch/timeline.html](https://forschung.stadtgeschichtebasel.ch/timeline.html)
2. see first row

### Anything else?

It may have to do with the inability of timeline to handle all edtf data. possibly especially BC dates

### PR DESCRIPTION
**Pull request by koilebeit**
# Pull request

## Proposed changes

fixes #165
## Types of changes

- [ ] New feature (non-breaking change which adds functionality).
- [ ] Enhancement (non-breaking change which enhances functionality)
- [x] Bug Fix (non-breaking change which fixes an issue).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist

- [x] I have read the **[CONTRIBUTING](./CONTRIBUTING.md)** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by CodeRabbit

- **New Features**
	- Enhanced date format handling and year navigation in the timeline layout.
	- Introduced a new method for filtering items by year based on regex patterns.

- **Bug Fixes**
	- Improved error handling for invalid regex patterns to ensure robust performance.

- **Refactor**
	- Updated sorting logic for date strings to ensure accurate processing and display, including new helper methods for better date management.

### CODE REVIEW

File: _plugins/sort_edtf.rb [hunk 1]
```diff
@@ -1,10 +1,47 @@
...
-            sorted.sort_by { |str| str[/\d+/].to_i }
+  module SortEDTF
+    def sort_edtf(array_of_strings)
+
+      # Remove any empty strings after stripping whitespace
+      valid_dates = array_of_strings.reject { |str| str.strip.empty? }
+
+      # Parse each string into a hash with numeric, original, and display representations
+      parsed_dates = valid_dates.map do |str|
+        cleaned_str = str.gsub(/[Xu]/, '0') # Replace X/x with 0 for numeric comparison
```
[the remark this answers is not in the document]
Comment (koilebeit) [reply _plugins/sort_edtf.rb:10]: it should not replace 'x'. this was a wrong description

File: _plugins/sort_edtf.rb [hunk 2]
```diff
@@ -1,10 +1,46 @@
...
-        def sort_edtf(array_of_strings)
-            sorted = array_of_strings.map { |str| str.gsub('X', '0') }
-            sorted.sort_by { |str| str[/\d+/].to_i }
-        end
+  module SortEDTF
+    def clean_date_string(str)
+      str.gsub(/[Xu]/, '0')
     end
+
+    def compute_numeric_value(cleaned_str)
```
Comment (maehr) [inline _plugins/sort_edtf.rb:7]: @koilebeit Minor remark: the function does not say clearly what it does (and the input is also not clearly labeled)

File: _plugins/sort_edtf.rb [hunk 3]
```diff
@@ -1,10 +1,46 @@
 module Jekyll
-    module SortEDTF
-        def sort_edtf(array_of_strings)
-            sorted = array_of_strings.map { |str| str.gsub('X', '0') }
-            sorted.sort_by { |str| str[/\d+/].to_i }
-        end
+  module SortEDTF
+    def clean_date_string(str)
```
Comment (maehr) [inline _plugins/sort_edtf.rb:3]: what about `def replace_x_with_parsable_0(str)`

File: _plugins/sort_edtf.rb [hunk 4]
```diff
@@ -1,10 +1,46 @@
 module Jekyll
-    module SortEDTF
-        def sort_edtf(array_of_strings)
-            sorted = array_of_strings.map { |str| str.gsub('X', '0') }
-            sorted.sort_by { |str| str[/\d+/].to_i }
-        end
+  module SortEDTF
```
Comment (maehr) [inline _plugins/sort_edtf.rb:2]: @koilebeit Think about the scope of your functions. Make everything private that is not exposed (and should not be used outside of this module) see https://ruby-for-beginners.rubymonstas.org/advanced/private_methods.html

**Review by maehr — CHANGES_REQUESTED** [review]
Thanks again!

**Comment by maehr** [comment]
I think jekyll is depending on the othe PR #172. We should try to merge that first

**Review by maehr — APPROVED** [review]
Thanks a bunch!

**Pull request merged**